### PR TITLE
[CI] enable operator benchmark on CPU

### PR DESCRIPTION
**Pull request by DiweiSun**
This is to enable operator benchmark for CPU to track op level performance. This PR is motivated by PR: https://github.com/pytorch/pytorch/issues/120982 and investigate feasibility in https://github.com/pytorch/pytorch/pull/127216

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mcarilli @ptrblck @leslie-fang-intel @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @xmfan

### CODE REVIEW

File: .ci/pytorch/test.sh [hunk 1]
```diff
@@ -1251,6 +1262,8 @@ elif [[ "$TEST_CONFIG" == distributed ]]; then
   if [[ "${SHARD_NUMBER}" == 1 ]]; then
     test_rpc
   fi
+elif [[ "${TEST_CONFIG}" == *cpu_operator_benchmark* ]]; then
+  test_operator_benchmark
```
Comment (chuanqi129) [inline .ci/pytorch/test.sh:1266]: Please add device as a parameter for this function

File: .ci/pytorch/test.sh [hunk 2]
```diff
@@ -1225,6 +1225,17 @@ test_linux_aarch64(){
...
 }
 
+test_operator_benchmark() {
+  TEST_REPORTS_DIR=$(pwd)/test/test-reports
+  TEST_DIR=$(pwd)
+  end_core=$(( CORES-1 ))
+  cd benchmarks/operator_benchmark/pt_extension
+  python setup.py install
+  cd "${TEST_DIR}"
+  taskset -c 0-"$end_core" python -m benchmarks/operator_benchmark/benchmark_all_test.py --output "${TEST_REPORTS_DIR}/operator_benchmark.csv"
```
Comment (chuanqi129) [inline .ci/pytorch/test.sh:1235]: Do we need to test all test cases? all just pick some subsets? And let's think about how to check the performance regression in this function

File: .github/workflows/operator_benchmark.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
+name: operator_benchmark
+
+on:
```
Comment (chuanqi129) [inline .github/workflows/operator_benchmark.yml:3]: TODO: change the trigger condition after the function ready

File: benchmarks/operator_benchmark/check_perf_csv.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        expected_perf = get_field(expected_csv, case, "Execution Time")
+
+        if expected_perf is None:
+            status = "FAIL"
+            print(
+            f"\n{case:34}  {status:9} Not Found. if it is expected, you can update in {expected_filename} to reflect the new module. "
+        )
+            continue
+
+        speed_up = perf / expected_perf - 1
```
Comment (chuanqi129) [inline benchmarks/operator_benchmark/check_perf_csv.py:31]:
```suggestion
        speed_up = expected_perf / perf
```

File: .github/workflows/operator_benchmark.yml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  workflow_dispatch:
+    inputs:
+      test_mode:
+        required: false
+        type: string
+        default: 'short'
+        description: tag filter for operator benchmarks, options from long, short, all
+
+concurrency:
+  group: group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref_name }}-${{ github.ref_type == 'branch' && github.sha }}-${{ github.event_name == 'workflow_dispatch' }}-${{ github.event_name == 'schedule' }}
```
Comment (chuanqi129) [inline .github/workflows/operator_benchmark.yml:20]: There are 2 `group:` in here

File: .github/workflows/operator_benchmark.yml [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    with:
+      build-environment: linux-jammy-py3.8-gcc11-build
+      docker-image-name: pytorch-linux-jammy-py3.8-gcc11-opbenchmark-build
+      test-matrix: |
+        { include: [
+          { config: "cpu_operator_benchmark_short", shard: 1, num_shards: 1, runner: "linux.12xlarge" },
+        ]}
+
+  linux-jammy-cpu-py3_8-gcc11-opbenchmark-on-demand-build:
+    if: inputs.test_mode
```
Comment (chuanqi129) [inline .github/workflows/operator_benchmark.yml:38]: it's better to use workflow_dispatch event as the condition

File: benchmarks/operator_benchmark/pt/nan_to_num_test.py [hunk 7]
```diff
@@ -43,6 +43,9 @@ def init(self, M, N, dtype, replace_inf, op_func):
         self.op_func = op_func
         self.set_module_name("nan_to_num")
 
+        if op_func is torch.nan_to_num_:
```
Comment (chuanqi129) [inline benchmarks/operator_benchmark/pt/nan_to_num_test.py:46]: Let's add comments to explain why we need this change

File: .ci/pytorch/test.sh [hunk 8]
```diff
@@ -1400,6 +1400,28 @@ test_linux_aarch64(){
...
+  mkdir -p "$TEST_REPORTS_DIR"
+  TEST_DIR=$(pwd)
+  CORES=$(lscpu | grep Core | awk '{print $4}')
+  end_core=$(( CORES-1 ))
+
+  cd benchmarks/operator_benchmark/pt_extension
+  python setup.py install
+
+  cd "${TEST_DIR}"/benchmarks/operator_benchmark
+  taskset -c 0-"$end_core" python -m benchmark_all_test --device $1 --output-dir "${TEST_REPORTS_DIR}/operator_benchmark_eager_float32_cpu.csv"
```
Comment (chuanqi129) [inline .ci/pytorch/test.sh:1414]: seems this function can't support long mode? We need a parameter for mode in here.

File: .ci/pytorch/test.sh [hunk 9]
```diff
@@ -1426,6 +1448,10 @@ elif [[ "$TEST_CONFIG" == distributed ]]; then
   if [[ "${SHARD_NUMBER}" == 1 ]]; then
     test_rpc
   fi
+elif [[ "${TEST_CONFIG}" == *operator_benchmark* ]]; then
+  if [[ "${TEST_CONFIG}" == *cpu* ]]; then
+    test_operator_benchmark cpu
```
Comment (chuanqi129) [inline .ci/pytorch/test.sh:1453]: also need parse the test mode in here and pass to the test function

**Review by leslie-fang-intel — APPROVED** [review]

File: .github/workflows/operator_benchmark.yml [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  linux-jammy-cpu-py3_9-gcc11-opbenchmark-on-demand-build:
+    if: ${{ github.event_name == 'workflow_dispatch' && github.repository_owner == 'pytorch' }}
+    name: linux-jammy-cpu-py3.9-gcc11-opbenchmark
+    uses: ./.github/workflows/_linux-build.yml
+    with:
+      build-environment: linux-jammy-py3.9-gcc11-build
+      docker-image-name: pytorch-linux-jammy-py3.9-gcc11-opbenchmark-build
+      test-matrix: |
+        { include: [
+          { config: "cpu_operator_benchmark_${{ inputs.test_mode }}", shard: 1, num_shards: 1, runner: "linux.12xlarge" },
```
Comment (sanchitintel) [inline .github/workflows/operator_benchmark.yml:47]: Hi @DiweiSun @chuanqi129, `linux.12xlarge` seems to use AWS `c5.12xlarge`, which is a Cascade Lake (Xeon Gen 2) instance.
Do we intend to check regressions on older Xeon generations through this PR? Thanks!

BTW, do op-level benchmarks have complete op coverage & dtype coverage? They didn't use to have full coverage last year, and I tried adding more, but abandoned plans due to CI costs concerns. Besides, Xeon 2nd gen doesn't support all ISAs, and instances running on Xeon 4th gen machines are more expensive.

**Comment by DiweiSun** [comment]
closed due to too much unexpected CCs. will be tracked with new PR: [[CI] enable operator benchmark on CPU by LifengWang · Pull Request #143733 · pytorch/pytorch](https://github.com/pytorch/pytorch/pull/143733)

**Pull request closed without merging**